src/util/sighash.rs

```diff
@@ -0,0 +1,909 @@
+//
+
+//! # Generalized, efficient, signature hash implementation
+//!
+//! Implementation of the algorithm to compute the message to be signed according to [Bip341](https://github.com/bitcoin/bips/blob/150ab6f5c3aca9da05fccc5b435e9667853407f4/bip-0341.mediawiki),
+//! [Bip143](https://github.com/bitcoin/bips/blob/99701f68a88ce33b2d0838eb84e115cef505b4c2/bip-0143.mediawiki)
+//! and legacy (before Bip143)
+//!
+
+pub use blockdata::transaction::SigHashType as LegacySigHashType;
+use consensus::{encode, Encodable};
+use core::fmt;
+use core::ops::{Deref, DerefMut};
+use hashes::{sha256, sha256d, Hash};
+use io;
+use util::taproot::{TapLeafHash, TapSighashHash};
+use SigHash;
+use {Script, Transaction, TxOut};
+
+use prelude::*;
+
+/// Efficiently calculates signature hash message for legacy, segwit and taproot inputs.
+#[derive(Debug)]
+pub struct SigHashCache<T: Deref<Target = Transaction>> {
+    /// Access to transaction required for various introspection, moreover type
+    /// `T: Deref<Target=Transaction>` allows to accept borrow and mutable borrow, the
+    /// latter in particular is necessary for [`SigHashCache::witness_mut`]
+    tx: T,
+
+    /// Common cache for taproot and segwit inputs. It's an option because it's not needed for legacy inputs
+    common_cache: Option<CommonCache>,
+
+    /// Cache for segwit v0 inputs, it's the result of another round of sha256 on `common_cache`
+    segwit_cache: Option<SegwitCache>,
+
+    /// Cache for taproot v1 inputs
+    taproot_cache: Option<TaprootCache>,
+}
+
+/// Values cached common between segwit and taproot inputs
+#[derive(Debug)]
+struct CommonCache {
+    prevouts: sha256::Hash,
+    sequences: sha256::Hash,
+
+    /// in theory, `outputs` could be `Option` since `NONE` and `SINGLE` doesn't need it, but since
+    /// `ALL` is the mostly used variant by large, we don't bother
+    outputs: sha256::Hash,
+}
+
+/// Values cached for segwit inputs, it's equal to [`CommonCache`] plus another round of `sha256`
+#[derive(Debug)]
+struct SegwitCache {
+    prevouts: sha256d::Hash,
+    sequences: sha256d::Hash,
+    outputs: sha256d::Hash,
+}
+
+/// Values cached for taproot inputs
+#[derive(Debug)]
+struct TaprootCache {
+    amounts: sha256::Hash,
+    script_pubkeys: sha256::Hash,
+}
+
+/// Contains outputs of previous transactions.
+/// In the case [`SigHashType`] variant is `ANYONECANPAY`, [`Prevouts::One`] may be provided
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum Prevouts<'u> {
+    /// `One` variant allows to provide the single Prevout needed. It's useful for example
+    /// when modifier `ANYONECANPAY` is provided, only prevout of the current input is needed.
+    /// The first `usize` argument is the input index this [`TxOut`] is referring to.
+    One(usize, &'u TxOut),
+    /// When `ANYONECANPAY` is not provided, or the caller is handy giving all prevouts so the same
+    /// variable can be used for multiple inputs.
+    All(&'u [TxOut]),
```

Comment:
nit: a slice is the wrong type, we want a reference to the underlying storage to ensure we have the whole slice, not just partial.

Reply:
I strongly disagree

Reply:
Another way to make this API safe would be via a smart constructor. E.g.

```rust
pub enum Prevouts<'u> {
//...
All(SafeSliceAll<'u>)
}
use private::SafeSliceAll;
mod private {
// self.0 not pub so can only be constructed locally.
pub struct SafeSliceAll<'u>(&'u [TxOut]);
impl<'u> From<&'u Vec<TxOut>> for SafeSliceAll<'u> {
//....
}
}
```

Allowing to construct an All from a arbitrary slice is error prone IMO since you can't guarantee it's *actually* all.

Reply:
we need to use such a smart constructor type because pub enums require all fields of variants are pub iirc.

Reply:
I don't understand why the user would be required to have the txouts all in a `Vec` like this.